File: backend/Services/ResumeTextExtractor.cs
using System.Diagnostics;
using System.Text;
using DocumentFormat.OpenXml.Packaging;
using UglyToad.PdfPig;

namespace OperationNexus.Api.Services;

public class ResumeTextExtractor : IResumeTextExtractor
{
    private static readonly string TesseractPath = FindBinary("tesseract");
    private static readonly string PdftoppmPath = FindBinary("pdftoppm");

    public string ExtractText(byte[] fileBytes, string filename)
    {
        var ext = Path.GetExtension(filename)?.ToLowerInvariant();
        var raw = ext switch
        {
            ".pdf" => ExtractFromPdf(fileBytes),
            ".docx" => ExtractFromDocx(fileBytes),
            ".doc" => ExtractFromDoc(fileBytes),
            _ => throw new NotSupportedException($"Unsupported resume format: {ext}")
        };
        return raw.Replace("\0", string.Empty);
    }

    private string ExtractFromPdf(byte[] fileBytes)
    {
        using var document = PdfDocument.Open(fileBytes);
        var sb = new StringBuilder();
        var hasImages = false;

        foreach (var page in document.GetPages())
        {
            var pageText = page.Text;
            if (!string.IsNullOrWhiteSpace(pageText))
                sb.AppendLine(pageText);

            if (page.GetImages().Any())
                hasImages = true;
        }

        var cleaned = sb.ToString().Replace("\0", string.Empty).Trim();
        var meaningfulChars = cleaned.Count(c => char.IsLetterOrDigit(c));

        if (meaningfulChars < 50 && !string.IsNullOrEmpty(PdftoppmPath) && !string.IsNullOrEmpty(TesseractPath))
        {
            Console.Error.WriteLine($"[TextExtractor] PDF has {document.NumberOfPages} page(s), {meaningfulChars} meaningful chars (hasImages={hasImages}) — attempting OCR via pdftoppm");
            var ocrText = ExtractWithOcr(fileBytes);
            if (!string.IsNullOrWhiteSpace(ocrText))
                return ocrText;
        }

        return sb.ToString();
    }

    private string ExtractWithOcr(byte[] pdfBytes)
    {
        if (string.IsNullOrEmpty(TesseractPath))
            throw new InvalidOperationException(
                "Tesseract CLI not found. Install via 'brew install tesseract' on macOS.");

        if (string.IsNullOrEmpty(PdftoppmPath))
            throw new InvalidOperationException(
                "pdftoppm not found. Install via 'brew install poppler' on macOS.");

        var tempDir = Path.Combine(Path.GetTempPath(), $"ocr-{Guid.NewGuid():N}");
        Directory.CreateDirectory(tempDir);

        try
        {
            var pdfPath = Path.Combine(tempDir, "input.pdf");
            File.WriteAllBytes(pdfPath, pdfBytes);

            var pagesPrefix = Path.Combine(tempDir, "page");
            var ppmPsi = new ProcessStartInfo
            {
                FileName = PdftoppmPath,
                ArgumentList = { "-png", "-r", "300", pdfPath, pagesPrefix },
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            };

            using (var ppmProcess = Process.Start(ppmPsi)!)
            {
                ppmProcess.WaitForExit(TimeSpan.FromSeconds(60));
                if (!ppmProcess.HasExited)
                {
                    ppmProcess.Kill(true);
                    throw new TimeoutException("pdftoppm timed out rendering PDF pages");
                }
                if (ppmProcess.ExitCode != 0)
                {
                    var stderr = ppmProcess.StandardError.ReadToEnd().Trim();
                    throw new InvalidOperationException($"pdftoppm failed (exit {ppmProcess.ExitCode}): {stderr}");
                }
            }

            var pageImages = Directory.GetFiles(tempDir, "page-*.png")
                .OrderBy(f => f)
                .ToArray();

            if (pageImages.Length == 0)
                throw new InvalidOperationException("pdftoppm produced no page images");

            Console.Error.WriteLine($"[OCR] Rendered {pageImages.Length} page image(s) at 300 DPI");

            var sb = new StringBuilder();
            foreach (var pageImage in pageImages)
            {
                var outputBase = pageImage.Replace(".png", "-ocr");

                var psi = new ProcessStartInfo
                {
                    FileName = TesseractPath,
                    ArgumentList = { pageImage, outputBase, "-l", "eng+spa" },
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    UseShellExecute = false,
                    CreateNoWindow = true,
                };

                using var process = Process.Start(psi)
                    ?? throw new InvalidOperationException("Failed to start Tesseract process");

                process.WaitForExit(TimeSpan.FromSeconds(30));

                if (!process.HasExited)
                {
                    process.Kill(true);
                    Console.Error.WriteLine($"[OCR] Tesseract timed out on {Path.GetFileName(pageImage)}");
                    continue;
                }

                var outputFile = outputBase + ".txt";
                if (process.ExitCode == 0 && File.Exists(outputFile))
                {
                    var text = File.ReadAllText(outputFile);
                    if (!string.IsNullOrWhiteSpace(text))
                        sb.AppendLine(text);
                }
                else
                {
                    var stderr = process.StandardError.ReadToEnd().Trim();
                    Console.Error.WriteLine($"[OCR] Tesseract failed on {Path.GetFileName(pageImage)}: {stderr}");
                }
            }

            Console.Error.WriteLine($"[OCR] Extracted {sb.Length} chars from {pageImages.Length} pages via pdftoppm + Tesseract");
            return sb.ToString();
        }
        catch (Exception ex) when (ex is not InvalidOperationException and not TimeoutException)
        {
            Console.Error.WriteLine($"[OCR] Pipeline failed: {ex}");
            throw new InvalidOperationException($"OCR extraction failed: {ex.Message}", ex);
        }
        finally
        {
            try { Directory.Delete(tempDir, true); } catch { }
        }
    }

    private static string ExtractFromDocx(byte[] fileBytes)
    {
        try
        {
            using var stream = new MemoryStream(fileBytes);
            using var doc = WordprocessingDocument.Open(stream, false);
            var body = doc?.MainDocumentPart?.Document?.Body;
            var text = body?.InnerText ?? string.Empty;
            if (!string.IsNullOrWhiteSpace(text))
                return text;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"[TextExtractor] OpenXml failed on .docx, falling back to textutil: {ex.Message}");
        }

        return ExtractWithTextutil(fileBytes, ".docx");
    }

    private static string ExtractFromDoc(byte[] fileBytes)
    {
        return ExtractWithTextutil(fileBytes, ".doc");
    }

    private static string ExtractWithTextutil(byte[] fileBytes, string extension)
    {
        var tempDir = Path.Combine(Path.GetTempPath(), $"textutil-{Guid.NewGuid():N}");
        Directory.CreateDirectory(tempDir);

        try
        {
            var inputPath = Path.Combine(tempDir, $"input{extension}");
            var outputPath = Path.Combine(tempDir, "output.txt");
            File.WriteAllBytes(inputPath, fileBytes);

            var psi = new ProcessStartInfo
            {
                FileName = "/usr/bin/textutil",
                ArgumentList = { "-convert", "txt", "-output", outputPath, inputPath },
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            };

            using var process = Process.Start(psi)
                ?? throw new InvalidOperationException("Failed to start textutil");

            process.WaitForExit(TimeSpan.FromSeconds(15));

            if (!process.HasExited)
            {
                process.Kill(true);
                throw new TimeoutException("textutil timed out converting document");
            }

            if (process.ExitCode != 0)
            {
                var stderr = process.StandardError.ReadToEnd().Trim();
                throw new InvalidOperationException($"textutil failed (exit {process.ExitCode}): {stderr}");
            }

            if (!File.Exists(outputPath))
                throw new InvalidOperationException("textutil produced no output");

            var text = File.ReadAllText(outputPath);
            Console.Error.WriteLine($"[TextExtractor] Extracted {text.Length} chars from {extension} via textutil");
            return text;
        }
        finally
        {
            try { Directory.Delete(tempDir, true); } catch { }
        }
    }

    private static string FindBinary(string name)
    {
        string[] searchDirs =
        [
            "/opt/homebrew/bin",
            "/usr/local/bin",
            "/usr/bin",
        ];

        foreach (var dir in searchDirs)
        {
            var path = Path.Combine(dir, name);
            if (File.Exists(path)) return path;
        }

        return string.Empty;
    }
}

File: src/main/services/resumeTextExtractor.ts
if (typeof globalThis.DOMMatrix === 'undefined') {
  globalThis.DOMMatrix = class DOMMatrix {
    a: number; b: number; c: number; d: number; e: number; f: number
    m11: number; m12: number; m13: number; m14: number
    m21: number; m22: number; m23: number; m24: number
    m31: number; m32: number; m33: number; m34: number
    m41: number; m42: number; m43: number; m44: number
    is2D = true
    isIdentity = true
    constructor(init?: string | number[]) {
      const v = Array.isArray(init) ? init : []
      if (v.length === 6) {
        this.a = v[0]; this.b = v[1]; this.c = v[2]; this.d = v[3]; this.e = v[4]; this.f = v[5]
      } else {
        this.a = 1; this.b = 0; this.c = 0; this.d = 1; this.e = 0; this.f = 0
      }
      this.m11 = this.a; this.m12 = this.b; this.m13 = 0; this.m14 = 0
      this.m21 = this.c; this.m22 = this.d; this.m23 = 0; this.m24 = 0
      this.m31 = 0; this.m32 = 0; this.m33 = 1; this.m34 = 0
      this.m41 = this.e; this.m42 = this.f; this.m43 = 0; this.m44 = 1
      this.is2D = true
      this.isIdentity = this.a === 1 && this.b === 0 && this.c === 0
        && this.d === 1 && this.e === 0 && this.f === 0
    }
    inverse() { return new DOMMatrix() }
    multiply() { return new DOMMatrix() }
    translate() { return new DOMMatrix() }
    scale() { return new DOMMatrix() }
    rotate() { return new DOMMatrix() }
    transformPoint(p: { x?: number; y?: number } = {}) {
      return { x: p.x ?? 0, y: p.y ?? 0, z: 0, w: 1 }
    }
    static fromMatrix() { return new DOMMatrix() }
    static fromFloat32Array() { return new DOMMatrix() }
    static fromFloat64Array() { return new DOMMatrix() }
  } as unknown as typeof globalThis.DOMMatrix
}

import { existsSync } from 'fs'
import { join } from 'path'
import { execFile as execFileCb } from 'child_process'
import { promisify } from 'util'
import { tmpdir } from 'os'
import { writeFileSync, readFileSync, mkdirSync, rmSync, readdirSync } from 'fs'

const execFile = promisify(execFileCb)
import { createLogger } from './logger'

const log = createLogger('ResumeTextExtractor')
import { randomUUID } from 'crypto'

const TESSERACT_PATH = findBinary('tesseract')
const PDFTOPPM_PATH = findBinary('pdftoppm')

function findBinary(name: string): string {
  const searchDirs = ['/opt/homebrew/bin', '/usr/local/bin', '/usr/bin']
  for (const dir of searchDirs) {
    const path = join(dir, name)
    if (existsSync(path)) return path
  }
  return ''
}

export const resumeTextExtractor = {
  async extractText(fileBytes: Buffer, filename: string): Promise<string> {
    const ext = filename.toLowerCase().split('.').pop() ?? ''
    let raw: string

    switch (`.${ext}`) {
      case '.pdf':
        raw = await extractFromPdf(fileBytes)
        break
      case '.docx':
        raw = await extractFromDocx(fileBytes)
        break
      case '.doc':
        raw = await extractWithTextutil(fileBytes, '.doc')
        break
      case '.jpg':
      case '.jpeg':
      case '.png':
        raw = await extractFromImage(fileBytes, `.${ext}`)
        break
      default:
        throw new Error(`Unsupported resume format: .${ext}`)
    }

    return raw.replace(/\0/g, '')
  },
}

async function extractFromPdf(fileBytes: Buffer): Promise<string> {
  try {
    const pdfjs = await import('pdfjs-dist')
    const data = new Uint8Array(fileBytes)
    const doc = await pdfjs.getDocument({ data, useSystemFonts: true }).promise
    const pages: string[] = []

    for (let i = 1; i <= doc.numPages; i++) {
      const page = await doc.getPage(i)
      const textContent = await page.getTextContent()
      const pageText = textContent.items
        .filter(item => 'str' in item)
        .map(item => (item as { str: string }).str)
        .join(' ')
      if (pageText.trim()) pages.push(pageText)
    }

    const fullText = pages.join('\n')
    const meaningfulChars = fullText.replace(/[^a-zA-Z0-9]/g, '').length

    if (meaningfulChars < 50 && PDFTOPPM_PATH && TESSERACT_PATH) {
      try {
        const ocrText = await extractWithOcr(fileBytes)
        if (ocrText.trim()) return ocrText
      } catch (ocrErr) {
        log.warn('OCR fallback failed, using pdfjs text', {
          meaningfulChars,
          error: ocrErr instanceof Error ? ocrErr.message : String(ocrErr),
        })
      }
    }

    return fullText
  } catch (err) {
    if (PDFTOPPM_PATH && TESSERACT_PATH) {
      try {
        const ocrText = await extractWithOcr(fileBytes)
        if (ocrText.trim()) return ocrText
      } catch (ocrErr) {
        log.warn('OCR last-resort also failed', {
          error: ocrErr instanceof Error ? ocrErr.message : String(ocrErr),
        })
      }
    }
    throw err
  }
}

async function extractFromDocx(fileBytes: Buffer): Promise<string> {
  try {
    const mammoth = await import('mammoth')
    const result = await mammoth.extractRawText({ buffer: fileBytes })
    if (result.value.trim()) return result.value
  } catch (err) {
    log.error('DOCX extraction with mammoth failed, falling back to textutil', err instanceof Error ? err : new Error(String(err)))
  }

  return await extractWithTextutil(fileBytes, '.docx')
}

async function extractWithOcr(pdfBytes: Buffer): Promise<string> {
  if (!TESSERACT_PATH) throw new Error('Tesseract CLI not found. Install via "brew install tesseract" on macOS.')
  if (!PDFTOPPM_PATH) throw new Error('pdftoppm not found. Install via "brew install poppler" on macOS.')

  const tempDir = join(tmpdir(), `ocr-${randomUUID().replace(/-/g, '')}`)
  mkdirSync(tempDir, { recursive: true })

  try {
    const pdfPath = join(tempDir, 'input.pdf')
    const pagesPrefix = join(tempDir, 'page')
    writeFileSync(pdfPath, pdfBytes)

    await execFile(PDFTOPPM_PATH, ['-png', '-r', '300', pdfPath, pagesPrefix], { timeout: 60000 })

    const pageImages = readdirSync(tempDir)
      .filter(f => f.startsWith('page-') && f.endsWith('.png'))
      .sort()

    if (pageImages.length === 0) throw new Error('pdftoppm produced no page images')

    const texts: string[] = []
    for (const pageImage of pageImages) {
      const imagePath = join(tempDir, pageImage)
      const outputBase = imagePath.replace('.png', '-ocr')

      try {
        await execFile(TESSERACT_PATH, [imagePath, outputBase, '-l', 'eng+spa'], { timeout: 30000 })
        const outputFile = `${outputBase}.txt`
        if (existsSync(outputFile)) {
          const text = readFileSync(outputFile, 'utf-8')
          if (text.trim()) texts.push(text)
        }
      } catch (err) {
        log.error(`OCR failed for page image ${pageImage}`, err instanceof Error ? err : new Error(String(err)))
        continue
      }
    }

    return texts.join('\n')
  } finally {
    try {
      rmSync(tempDir, { recursive: true, force: true })
    } catch (err) {
      log.error('Failed to cleanup OCR temp directory', err instanceof Error ? err : new Error(String(err)))
    }
  }
}

async function extractFromImage(imageBytes: Buffer, extension: string): Promise<string> {
  if (!TESSERACT_PATH) throw new Error('Tesseract CLI not found. Install via "brew install tesseract" on macOS.')

  const tempDir = join(tmpdir(), `ocr-img-${randomUUID().replace(/-/g, '')}`)
  mkdirSync(tempDir, { recursive: true })

  try {
    const inputPath = join(tempDir, `input${extension}`)
    const outputBase = join(tempDir, 'output')
    writeFileSync(inputPath, imageBytes)

    await execFile(TESSERACT_PATH, [inputPath, outputBase, '-l', 'eng+spa'], { timeout: 30000 })

    const outputFile = `${outputBase}.txt`
    if (existsSync(outputFile)) {
      return readFileSync(outputFile, 'utf-8')
    }

    throw new Error('Tesseract produced no output')
  } finally {
    try {
      rmSync(tempDir, { recursive: true, force: true })
    } catch (err) {
      log.error('Failed to cleanup image OCR temp directory', err instanceof Error ? err : new Error(String(err)))
    }
  }
}

async function extractWithTextutil(fileBytes: Buffer, extension: string): Promise<string> {
  const tempDir = join(tmpdir(), `textutil-${randomUUID().replace(/-/g, '')}`)
  mkdirSync(tempDir, { recursive: true })

  try {
    const inputPath = join(tempDir, `input${extension}`)
    const outputPath = join(tempDir, 'output.txt')
    writeFileSync(inputPath, fileBytes)

    await execFile('/usr/bin/textutil', ['-convert', 'txt', '-output', outputPath, inputPath], { timeout: 15000 })

    if (!existsSync(outputPath)) throw new Error('textutil produced no output')
    return readFileSync(outputPath, 'utf-8')
  } finally {
    try {
      rmSync(tempDir, { recursive: true, force: true })
    } catch (err) {
      log.error('Failed to cleanup textutil temp directory', err instanceof Error ? err : new Error(String(err)))
    }
  }
}
